File: poly_apps/flutter_bloom/lib/common/i18n/i18n_service.dart
/// Centralized internationalization service for Flutter Bloom
/// Provides multi-language support with reactive updates
library i18n_service;

import 'package:flutter/material.dart';
import 'translations.dart';

class I18nService extends ChangeNotifier {
  static final I18nService _instance = I18nService._internal();
  factory I18nService() => _instance;
  I18nService._internal();

  Locale _currentLocale = const Locale('zh', 'CN');

  Locale get currentLocale => _currentLocale;
  String get languageCode => _currentLocale.languageCode;

  /// Get translation by key
  String translate(String key, {Map<String, dynamic>? params}) {
    final translations = _getTranslations();
    String value = _getNestedValue(translations, key) ?? key;

    if (params != null) {
      params.forEach((paramKey, paramValue) {
        value = value.replaceAll('{$paramKey}', paramValue.toString());
      });
    }

    return value;
  }

  /// Get nested translation value
  String? _getNestedValue(Map<String, dynamic> map, String key) {
    final keys = key.split('.');
    dynamic current = map;

    for (final k in keys) {
      if (current is Map<String, dynamic> && current.containsKey(k)) {
        current = current[k];
      } else {
        return null;
      }
    }

    return current?.toString();
  }

  /// Get current translations
  Map<String, dynamic> _getTranslations() {
    switch (_currentLocale.languageCode) {
      case 'en':
        return Translations.en;
      case 'zh':
      default:
        return Translations.zhCN;
    }
  }

  /// Change language
  Future<void> changeLanguage(String languageCode) async {
    Locale newLocale;

    switch (languageCode) {
      case 'en':
        newLocale = const Locale('en', 'US');
        break;
      case 'zh':
      default:
        newLocale = const Locale('zh', 'CN');
        break;
    }

    if (_currentLocale != newLocale) {
      _currentLocale = newLocale;
      notifyListeners();
    }
  }

  /// Get supported locales
  List<Locale> get supportedLocales => [
    const Locale('zh', 'CN'),
    const Locale('en', 'US'),
  ];
}

/// Extension for easy translation access
extension I18nString on String {
  String get tr => I18nService().translate(this);
  String trParams(Map<String, dynamic> params) => I18nService().translate(this, params: params);
}

File: poly_apps/appfactory-master-dashboard/services/i18nService.ts
/**
 * i18n Internationalization Service
 * Provides multi-language support for the application
 */
import { en, TranslationKeys } from '../locales/en';
import { zh } from '../locales/zh';
import { ja } from '../locales/ja';

// Supported languages list
export const SUPPORTED_LANGUAGES = {
  en: 'English',
  zh: '中文',
  ja: '日本語',
} as const;

export type SupportedLanguage = keyof typeof SUPPORTED_LANGUAGES;

// Language resource mapping
const translations: Record<SupportedLanguage, TranslationKeys> = {
  en,
  zh,
  ja,
};

// Helper function to get nested property
function getNestedProperty(obj: Record<string, unknown>, path: string): string {
  const keys = path.split('.');
  let result = obj;

  for (const key of keys) {
    if (result && typeof result === 'object' && key in result) {
      result = result[key];
    } else {
      return path; // If not found, return original path
    }
  }

  return typeof result === 'string' ? result : path;
}

/**
 * i18n Service Class
 * Manages language switching and translation
 */
class I18nService {
  private currentLanguage: SupportedLanguage = 'zh';
  private listeners: Set<() => void> = new Set();

  /**
   * Get current language
   */
  getCurrentLanguage(): SupportedLanguage {
    return this.currentLanguage;
  }

  /**
   * Set current language
   */
  setLanguage(language: SupportedLanguage): void {
    if (this.currentLanguage !== language) {
      this.currentLanguage = language;
      this.notifyListeners();
    }
  }

  /**
   * Translate text
   * @param key - Translation key, supports dot-separated nested paths, e.g. 'dashboard.title'
   * @param params - Optional parameters object for placeholder replacement
   */
  t(key: string, params?: Record<string, string | number>): string {
    const translation = getNestedProperty(translations[this.currentLanguage], key);

    // If params exist, replace placeholders
    if (params) {
      return Object.entries(params).reduce((text, [paramKey, paramValue]) => {
        return text.replace(new RegExp(`{${paramKey}}`, 'g'), String(paramValue));
      }, translation);
    }

    return translation;
  }

  /**
   * Add language change listener
   */
  addListener(listener: () => void): void {
    this.listeners.add(listener);
  }

  /**
   * Remove language change listener
   */
  removeListener(listener: () => void): void {
    this.listeners.delete(listener);
  }

  /**
   * Notify all listeners
   */
  private notifyListeners(): void {
    this.listeners.forEach(listener => listener());
  }

  /**
   * Get all supported languages
   */
  getSupportedLanguages(): typeof SUPPORTED_LANGUAGES {
    return SUPPORTED_LANGUAGES;
  }
}

// Export singleton
export const i18nService = new I18nService();
